Fix: Handle empty func_name and None values in GLM MoE detectors

Leoyzen opened the pull request:
- Fix AssertionError when func_name is empty by adding proper checks
- Handle None values before calling strip() method to prevent AttributeError
- Add boundary case handling for partial matches in GLM-4.7 and GLM-4 MoE detectors
- Add test cases to verify the fixes for empty function name handling

<!-- Thank you for your contribution! Please follow these guidelines to enhance your pull request. If anything is unclear, submit your PR and reach out to maintainers for assistance. Join our Slack community at https://slack.sglang.io to discuss further. -->

## Motivation

This pull request addresses critical issues in the GLM-4.7 and GLM-4 MoE function call parsers that were causing runtime errors. Specifically, the parsers were throwing AssertionError when encountering empty function names and AttributeError when trying to call strip() on None values. These issues occurred when the model outputted only the start token without a complete function name, which is a common occurrence during streaming inference. The fixes ensure robust handling of partial and incomplete function call patterns, improving the reliability of function call parsing for GLM models.

See also 

* https://github.com/sgl-project/sglang/issues/15721

## Modifications

1. **GLM-4.7 MoE Detector** (`python/sglang/srt/function_call/glm47_moe_detector.py`):
   - Added proper null checks before accessing regex match groups
   - Implemented safe handling of empty function names to prevent AssertionError
   - Added None checks before calling strip() method to prevent AttributeError
   - Improved boundary case handling for partial matches during streaming

2. **GLM-4 MoE Detector** (`python/sglang/srt/function_call/glm4_moe_detector.py`):
   - Applied similar fixes as GLM-4.7 to ensure consistency
   - Added safe checks for function name validation
   - Improved handling of None values in parameter processing

3. **Test Cases** (`test/registered/function_call/test_function_call_parser.py`):
   - Added `test_empty_function_name_handling` test cases for both detectors
   - Verified that the parsers handle partial function call patterns gracefully
   - Ensured that no exceptions are thrown when processing incomplete function names

## Accuracy Tests

No specific accuracy tests were required as this PR addresses robustness and error handling rather than changing the core functionality of the parsers. The existing test suite ensures that the functional behavior remains unchanged while improving error handling.

## Benchmarking and Profiling

No performance impact is expected from these changes as they only add null checks and safe handling of edge cases. The core parsing logic remains unchanged, ensuring the same performance characteristics.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).
- [x] Work with maintainers to merge your PR. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process)

Leoyzen committed:
Fix: Handle empty func_name and None values in GLM MoE detectors

- Fix AssertionError when func_name is empty by adding proper checks
- Handle None values before calling strip() method to prevent AttributeError
- Add boundary case handling for partial matches in GLM-4.7 and GLM-4 MoE detectors
- Add test cases to verify the fixes for empty function name handling

Leoyzen committed:
Merge branch 'main' into feature/yuchen.liu/glm47-parser-fix

Leoyzen committed:
fixup! Fix: Handle empty func_name and None values in GLM MoE detectors

Leoyzen committed:
[fix] remove duplicate tool response in glm4_moe_detector.

Leoyzen commented on python/sglang/srt/function_call/glm4_moe_detector.py:
```diff
@@ -447,7 +458,13 @@ def parse_streaming_increment(
                 # Send tool name first if not sent yet
                 if not self.current_tool_name_sent:
-                    assert func_name, "func_name should not be empty"
+                    calls.append(
```
@lambert0312 fixed.

Leoyzen committed:
Fix function name truncation in GLM47 MoE detector

- Fixed issue where function names were sent prematurely when incomplete
- Modified regex matching to ensure function name is only sent when complete
- Added checks for <arg_key> or end token before sending function name
- Fixed state management for multiple tool calls to properly reset current_tool_name_sent

This resolves the issue where 'ask_followup_question' was being truncated to 'ask'

Leoyzen commented:
Fix function name truncation in GLM47 MoE detector

## Summary
Fixed an issue in the GLM47 MoE detector where function names were being sent prematurely when incomplete during streaming, causing function name truncation. For example, 'ask_followup_question' was being truncated to 'ask'.

## Problem
In `parse_streaming_increment`, when streaming function calls character-by-character, the function name was being sent as soon as it was detected, even if it was incomplete. This happened because the detector was not waiting for the `<arg_key>` tag or the end token `</tool_call>` to confirm the function name was complete.

## Solution
1. Modified the regex matching logic to properly separate function name from argument tags
2. Added condition to only send function name when `<arg_key>` tag or end token `</tool_call>` is encountered
3. Fixed state management for multiple tool calls to properly reset `current_tool_name_sent` after each complete tool call

## Changes
- Updated regex patterns to correctly identify complete function names
- Added checks for `<arg_key>` or end token before sending function names
- Fixed state reset logic for handling multiple consecutive tool calls
- All existing tests continue to pass

## Verification
- Function names are now only sent when complete (e.g., 'ask_followup_question' instead of 'ask')
- Multiple tool call scenarios work correctly
- Streaming functionality preserved
- All GLM47 MoE detector tests pass (9/9)

Leoyzen committed:
fixup! Fix function name truncation in GLM47 MoE detector

Leoyzen committed:
fix: prevent duplicate empty objects in GLM-4.7 MOE detector

- Add _tool_call_completed flag to prevent duplicate processing of end tokens
- Ensure end marker processing logic only executes once per tool call
- Add tests to verify no duplicate empty objects for no-argument functions
- Fix potential issue where  would generate duplicate  arguments

JustinTong0323 reviewed (approved):
Verified and get 100% tool call accuracy, Great! Thanks for the contribution!

Leoyzen committed:
Fix duplicate empty object issue in GLM-4.7 MoE detector for no-arg function calls

- Ensure self.streamed_args_for_tool array is updated consistently with self._last_arguments in parse_streaming_increment method
- Add unit test to verify the fix, ensuring no-argument function calls do not produce duplicate {} outputs

JustinTong0323 commented:
/tag-and-rerun-ci

cynial commented:
@Leoyzen  Below is a simplified version of the prompt. It doesn't happen every time, but there's a certain probability of character omission occurring. I found it tends to occur in this situation: **Chinese punctuation marks like ：。etc., immediately followed by `<tool_call>`**

missing character
```
结构：
```

stream output
```
data: {"id":"d8f9223c2ab240229a7a446db6667474","object":"chat.completion.chunk","created":1766726708,"model":"glm-4.7","choices":[{"index":0,"delta":{"role":"assistant","content":"","reasoning_content":null,"tool_calls":null},"logprobs":null,"finish_reason":null,"matched_stop":null}],"usage":null}

data: {"id":"d8f9223c2ab240229a7a446db6667474","object":"chat.completion.chunk","created":1766726708,"model":"glm-4.7","choices":[{"index":0,"delta":{"role":null,"content":null,"reasoning_content":"好的","tool_calls":null},"logprobs":null,"finish_reason":null,"matched_stop":null}],"usage":null}

data: {"id":"d8f9223c2ab240229a7a446db6667474","object":"chat.completion.chunk","created":1766726708,"model":"glm-4.7","choices":[{"index":0,"delta":{"role":null,"content":null,"reasoning_content":"，用户","tool_calls":null},"logprobs":null,"finish_reason":null,"matched_stop":null}],"usage":null}

data: {"id":"d8f9223c2ab240229a7a446db6667474","object":"chat.completion.chunk","created":1766726708,"model":"glm-4.7","choices":[{"index":0,"delta":{"role":null,"content":null,"reasoning_content":"要求","tool_calls":null},"logprobs":null,"finish_reason":null,"matched_stop":null}],"usage":null}

data: {"id":"d8f9223c2ab240229a7a446db6667474","object":"chat.completion.chunk","created":1766726709,"model":"glm-4.7","choices":[{"index":0,"delta":{"role":null,"content":null,"reasoning_content":"查找","tool_calls":null},"logprobs":null,"finish_reason":null,"matched_stop":null}],"usage":null}

data: {"id":"d8f9223c2ab240229a7a446db6667474","object":"chat.completion.chunk","created":1766726709,"model":"glm-4.7","choices":[{"index":0,"delta":{"role":null,"content":null,"reasoning_content":"所有","tool_calls":null},"logprobs":null,"finish_reason":null,"matched_stop":null}],"usage":null}

data: {"id":"d8f9223c2ab240229a7a446db6667474","object":"chat.completion.chunk","created":1766726709,"model":"glm-4.7","choices":[{"index":0,"delta":{"role":null,"content":null,"reasoning_content":"相关的","tool_calls":null},"logprobs":null,"finish_reason":null,"matched_stop":null}],"usage":null}

data: {"id":"d8f9223c2ab240229a7a446db6667474","object":"chat.completion.chunk","created":1766726709,"model":"glm-4.7","choices":[{"index":0,"delta":{"role":null,"content":null,"reasoning_content":"配置文件并列","tool_calls":null},"logprobs":null,"finish_reason":null,"matched_stop":null}],"usage":null}

data: {"id":"d8f9223c2ab240229a7a446db6667474","object":"chat.completion.chunk","created":1766726709,"model":"glm-4.7","choices":[{"index":0,"delta":{"role":null,"content":null,"reasoning_content":"出它们的","tool_calls":null},"logprobs":null,"finish_reason":null,"matched_stop":null}],"usage":null}

data: {"id":"d8f9223c2ab240229a7a446db6667474","object":"chat.completion.chunk","created":1766726709,"model":"glm-4.7","choices":[{"index":0,"delta":{"role":null,"content":null,"reasoning_content":"路径。我需要","tool_calls":null},"logprobs":null,"finish_reason":null,"matched_stop":null}],"usage":null}

data: {"id":"d8f9223c2ab240229a7a446db6667474","object":"chat.completion.chunk","created":1766726709,"model":"glm-4.7","choices":[{"index":0,"delta":{"role":null,"content":null,"reasoning_content":"先查看","tool_calls":null},"logprobs":null,"finish_reason":null,"matched_stop":null}],"usage":null}

data: {"id":"d8f9223c2ab240229a7a446db6667474","object":"chat.completion.chunk","created":1766726709,"model":"glm-4.7","choices":[{"index":0,"delta":{"role":null,"content":null,"reasoning_content":"当前目录的结构，","tool_calls":null},"logprobs":null,"finish_reason":null,"matched_stop":null}],"usage":null}

data: {"id":"d8f9223c2ab240229a7a446db6667474","object":"chat.completion.chunk","created":1766726709,"model":"glm-4.7","choices":[{"index":0,"delta":{"role":null,"content":null,"reasoning_content":"看看","tool_calls":null},"logprobs":null,"finish_reason":null,"matched_stop":null}],"usage":null}

data: {"id":"d8f9223c2ab240229a7a446db6667474","object":"chat.completion.chunk","created":1766726709,"model":"glm-4.7","choices":[{"index":0,"delta":{"role":null,"content":null,"reasoning_content":"有哪些目录","tool_calls":null},"logprobs":null,"finish_reason":null,"matched_stop":null}],"usage":null}

data: {"id":"d8f9223c2ab240229a7a446db6667474","object":"chat.completion.chunk","created":1766726709,"model":"glm-4.7","choices":[{"index":0,"delta":{"role":null,"content":null,"reasoning_content":"可能","tool_calls":null},"logprobs":null,"finish_reason":null,"matched_stop":null}],"usage":null}

data: {"id":"d8f9223c2ab240229a7a446db6667474","object":"chat.completion.chunk","created":1766726709,"model":"glm-4.7","choices":[{"index":0,"delta":{"role":null,"content":null,"reasoning_content":"包含配置文件。\n\n","tool_calls":null},"logprobs":null,"finish_reason":null,"matched_stop":null}],"usage":null}

data: {"id":"d8f9223c2ab240229a7a446db6667474","object":"chat.completion.chunk","created":1766726709,"model":"glm-4.7","choices":[{"index":0,"delta":{"role":null,"content":null,"reasoning_content":"让我先从","tool_calls":null},"logprobs":null,"finish_reason":null,"matched_stop":null}],"usage":null}

data: {"id":"d8f9223c2ab240229a7a446db6667474","object":"chat.completion.chunk","created":1766726709,"model":"glm-4.7","choices":[{"index":0,"delta":{"role":null,"content":null,"reasoning_content":"根目录开始查看","tool_calls":null},"logprobs":null,"finish_reason":null,"matched_stop":null}],"usage":null}

data: {"id":"d8f9223c2ab240229a7a446db6667474","object":"chat.completion.chunk","created":1766726709,"model":"glm-4.7","choices":[{"index":0,"delta":{"role":null,"content":null,"reasoning_content":"。","tool_calls":null},"logprobs":null,"finish_reason":null,"matched_stop":null}],"usage":null}

data: {"id":"d8f9223c2ab240229a7a446db6667474","object":"chat.completion.chunk","created":1766726709,"model":"glm-4.7","choices":[{"index":0,"delta":{"role":null,"content":"我来","reasoning_content":null,"tool_calls":null},"logprobs":null,"finish_reason":null,"matched_stop":null}],"usage":null}

data: {"id":"d8f9223c2ab240229a7a446db6667474","object":"chat.completion.chunk","created":1766726709,"model":"glm-4.7","choices":[{"index":0,"delta":{"role":null,"content":"帮您查找所有","reasoning_content":null,"tool_calls":null},"logprobs":null,"finish_reason":null,"matched_stop":null}],"usage":null}

data: {"id":"d8f9223c2ab240229a7a446db6667474","object":"chat.completion.chunk","created":1766726709,"model":"glm-4.7","choices":[{"index":0,"delta":{"role":null,"content":"相关的配置文件。","reasoning_content":null,"tool_calls":null},"logprobs":null,"finish_reason":null,"matched_stop":null}],"usage":null}

data: {"id":"d8f9223c2ab240229a7a446db6667474","object":"chat.completion.chunk","created":1766726709,"model":"glm-4.7","choices":[{"index":0,"delta":{"role":null,"content":"首先让我查看","reasoning_content":null,"tool_calls":null},"logprobs":null,"finish_reason":null,"matched_stop":null}],"usage":null}

data: {"id":"d8f9223c2ab240229a7a446db6667474","object":"chat.completion.chunk","created":1766726709,"model":"glm-4.7","choices":[{"index":0,"delta":{"role":null,"content":"目录","reasoning_content":null,"tool_calls":null},"logprobs":null,"finish_reason":null,"matched_stop":null}],"usage":null}

data: {"id":"d8f9223c2ab240229a7a446db6667474","object":"chat.completion.chunk","created":1766726709,"model":"glm-4.7","choices":[{"index":0,"delta":{"role":null,"content":null,"reasoning_content":null,"tool_calls":[{"id":"call_4e8e2157fdcc4d7ab0c2bef0","index":0,"type":"function","function":{"name":"list_dir","arguments":""}}]},"logprobs":null,"finish_reason":null,"matched_stop":null}],"usage":null}

data: {"id":"d8f9223c2ab240229a7a446db6667474","object":"chat.completion.chunk","created":1766726709,"model":"glm-4.7","choices":[{"index":0,"delta":{"role":null,"content":null,"reasoning_content":null,"tool_calls":[{"id":null,"index":0,"type":"function","function":{"name":null,"arguments":"{\"path\": "}}]},"logprobs":null,"finish_reason":null,"matched_stop":null}],"usage":null}

data: {"id":"d8f9223c2ab240229a7a446db6667474","object":"chat.completion.chunk","created":1766726709,"model":"glm-4.7","choices":[{"index":0,"delta":{"role":null,"content":null,"reasoning_content":null,"tool_calls":[{"id":null,"index":0,"type":"function","function":{"name":null,"arguments":"\"/github/sg"}}]},"logprobs":null,"finish_reason":null,"matched_stop":null}],"usage":null}

data: {"id":"d8f9223c2ab240229a7a446db6667474","object":"chat.completion.chunk","created":1766726709,"model":"glm-4.7","choices":[{"index":0,"delta":{"role":null,"content":null,"reasoning_content":null,"tool_calls":[{"id":null,"index":0,"type":"function","function":{"name":null,"arguments":"lang\", "}}]},"logprobs":null,"finish_reason":null,"matched_stop":null}],"usage":null}

data: {"id":"d8f9223c2ab240229a7a446db6667474","object":"chat.completion.chunk","created":1766726709,"model":"glm-4.7","choices":[{"index":0,"delta":{"role":null,"content":null,"reasoning_content":null,"tool_calls":[{"id":null,"index":0,"type":"function","function":{"name":null,"arguments":"\"explanation\": "}}]},"logprobs":null,"finish_reason":null,"matched_stop":null}],"usage":null}

data: {"id":"d8f9223c2ab240229a7a446db6667474","object":"chat.completion.chunk","created":1766726709,"model":"glm-4.7","choices":[{"index":0,"delta":{"role":null,"content":null,"reasoning_content":null,"tool_calls":[{"id":null,"index":0,"type":"function","function":{"name":null,"arguments":"\"列出"}}]},"logprobs":null,"finish_reason":null,"matched_stop":null}],"usage":null}

data: {"id":"d8f9223c2ab240229a7a446db6667474","object":"chat.completion.chunk","created":1766726709,"model":"glm-4.7","choices":[{"index":0,"delta":{"role":null,"content":null,"reasoning_content":null,"tool_calls":[{"id":null,"index":0,"type":"function","function":{"name":null,"arguments":"当前"}}]},"logprobs":null,"finish_reason":null,"matched_stop":null}],"usage":null}

data: {"id":"d8f9223c2ab240229a7a446db6667474","object":"chat.completion.chunk","created":1766726709,"model":"glm-4.7","choices":[{"index":0,"delta":{"role":null,"content":null,"reasoning_content":null,"tool_calls":[{"id":null,"index":0,"type":"function","function":{"name":null,"arguments":"工作目录的内容"}}]},"logprobs":null,"finish_reason":null,"matched_stop":null}],"usage":null}

data: {"id":"d8f9223c2ab240229a7a446db6667474","object":"chat.completion.chunk","created":1766726709,"model":"glm-4.7","choices":[{"index":0,"delta":{"role":null,"content":null,"reasoning_content":null,"tool_calls":[{"id":null,"index":0,"type":"function","function":{"name":null,"arguments":"以"}}]},"logprobs":null,"finish_reason":null,"matched_stop":null}],"usage":null}

data: {"id":"d8f9223c2ab240229a7a446db6667474","object":"chat.completion.chunk","created":1766726709,"model":"glm-4.7","choices":[{"index":0,"delta":{"role":null,"content":null,"reasoning_content":null,"tool_calls":[{"id":null,"index":0,"type":"function","function":{"name":null,"arguments":"查找配置文件\""}}]},"logprobs":null,"finish_reason":null,"matched_stop":null}],"usage":null}

data: {"id":"d8f9223c2ab240229a7a446db6667474","object":"chat.completion.chunk","created":1766726709,"model":"glm-4.7","choices":[{"index":0,"delta":{"role":null,"content":null,"reasoning_content":null,"tool_calls":[{"id":null,"index":0,"type":"function","function":{"name":null,"arguments":"}"}}]},"logprobs":null,"finish_reason":null,"matched_stop":null}],"usage":null}

data: {"id":"d8f9223c2ab240229a7a446db6667474","object":"chat.completion.chunk","created":1766726709,"model":"glm-4.7","choices":[{"index":0,"delta":{"role":null,"content":null,"reasoning_content":null,"tool_calls":null},"logprobs":null,"finish_reason":"tool_calls","matched_stop":151338}],"usage":null}

data: {"id":"d8f9223c2ab240229a7a446db6667474","object":"chat.completion.chunk","created":1766726709,"model":"glm-4.7","choices":[],"usage":{"prompt_tokens":338,"total_tokens":421,"completion_tokens":83,"prompt_tokens_details":{"cached_tokens":336},"reasoning_tokens":0}}

data: [DONE]


```

server log
```
out={'text': '好的，用户要求查找所有相关的配置文件并列出它们的路径。我需要先查看当前目录的结构，看看有哪些目录可能包含配置文件。\n\n让我先从根目
录开始查看。</think>我来帮您查找所有相关的配置文件。首先让我查看目录结构：<tool_call>list_dir<arg_key>path</arg_key><arg_value>/github/sglang</arg_value><arg_key>explanation</arg_key><arg_value>列出当前工作目录的内容以查找配置文件</arg_value></tool_call>'
```

prompt

```
{
    "messages": [
        {
            "role": "assistant",
            "content": "configs 目录不存在，让我先看看当前工作目录下有哪些子目录：",
            "tool_calls": [
                {
                    "id": "call_ef99fabf965b400099a67093",
                    "type": "function",
                    "function": {
                        "name": "list_dir",
                        "arguments": "{\"explanation\": \"列出当前工作目录的子目录和文件，查找可能包含配置文件的目录\", \"path\": \"/github/sglang\"}"
                    }
                }
            ]
        }
    ],
    "model": "glm-4.7",
    "stream": true,
    "stream_options": {
        "include_usage": true
    },
    "temperature": 1,
    "tools": [
        {
            "type": "function",
            "function": {
                "name": "list_dir",
                "description": "List the contents of a directory. Directory path must be an absolute path to a directory that exists. If recursive is true, it will list all files and directories recursively. For each child in the directory, output will have: path to the directory, whether it is a directory or file, size in bytes and modified time if file, and number of children (recursive) if directory. Remember, this tool only return the first 50 items",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "explanation": {
                            "type": "string",
                            "description": "In one sentence explanation: Can we switch to using codebase_search for more efficient code retrieval?"
                        },
                        "path": {
                            "type": "string",
                            "description": "Path to list contents of, should be absolute path to a directory"
                        }
                    },
                    "required": [
                        "path"
                    ]
                }
            }
        }
    ],
    "top_p": 0.95
}
```

cynial commented:
sometimes, it directly outputs `没有输出，让我再试试其他可能存在的目录：<tool_call>list_dir<arg_key>explanation</arg_key><arg_value>检查根目录下有哪些可用目录</arg_value><arg_key>path</arg_key><arg_value>/</arg_value></tool_call>` as text, instead of making the tool call.

cynial commented:
@Leoyzen not sure, but it significantly affects the response quality.

Leoyzen committed:
[fix]Fix Chinese punctuation problems.

Leoyzen committed:
fixup! [fix]Fix Chinese punctuation problems.

JustinTong0323 committed:
cleanup test

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

Leoyzen committed:
fix(function_call): Fix tool_id increment logic in GLM-4.7 MoE streaming parser

Fix incorrect tool_id auto-increment in parse_streaming_increment() that caused
state management issues when processing Chinese punctuation followed by tool calls.

The problematic logic at lines 508-521 would increment current_tool_id when
detecting bot_token in new_text, even on first encounter. This caused:
- tool_id to jump from -1 → 0 (init) → 1 (incorrect increment)
- State confusion when receiving complete chunks with Chinese punctuation

Removed the premature auto-increment logic. Now tool_id is only incremented
after completing a tool call (line 671), ensuring correct state management.

All 24 existing tests pass, plus new test case added to prevent regression.

Leoyzen committed:
Optimize parse_streaming_increment method in Glm47MoeDetector

Leoyzen committed:
refactor(test): restructure Glm47MoeDetector tests based on MTP model behavior

- Reduce test count from 35 to 22 (-37%) by eliminating redundancies
- Add 3 critical MTP core scenario tests covering function/string splitting,
  no-arg functions, and number/JSON array handling
- Add 9 comprehensive boundary tests for edge cases
- Consolidate 6 Chinese punctuation tests into 1 generic test
- Remove 3 duplicate no-arg function tests
- Remove 5 duplicate function name truncation tests
- Add detailed English docstrings explaining scenario and purpose for each test
- All 22 tests passing with 100% success rate
- Maintain single TestGlm47MoeDetector class without additional helpers
- Document known limitation: text after tool calls not captured (TODO)

Leoyzen committed:
fix: preserve normal text after tool calls in Glm47MoeDetector

Previously, the detect_and_parse() method only extracted text before the
first <tool_call> tag, discarding any text appearing after tool calls.
This caused test_boundary_normal_text_mixed_with_tool to fail.

Now extracts all normal text segments (before, between, and after tool
calls) using regex iteration and concatenates them properly.

All 22 tests now pass.

Leoyzen committed:
Merge branch 'main' into feature/yuchen.liu/glm47-parser-fix

Leoyzen committed:
refactor: simplify get_argument_type() with cleaner dict access and reduced nesting

Leoyzen committed:
remove previous tests because of newly added delicated test file.

JustinTong0323 commented:
/tag-and-rerun-ci

JustinTong0323 committed:
register glm47 function call test to CI

Signed-off-by: Xinyuan Tong <xinyuantong.cs@gmail.com>

Leoyzen committed:
Merge branch 'main' into feature/yuchen.liu/glm47-parser-fix

JustinTong0323 committed:
Merge branch 'main' into feature/yuchen.liu/glm47-parser-fix

liaol commented on python/sglang/srt/function_call/glm47_moe_detector.py:
```diff
@@ -369,6 +411,176 @@ def _process_xml_to_json_streaming(
+        # Update state
+        self._last_arguments += json_increment
+        self.streamed_args_for_tool[self.current_tool_id] += json_increment
+        self._streamed_raw_length = current_raw_length
```
```self._streamed_raw_length = current_raw_length``` must be executed before：
```
if not json_increment:
    return None
 ```
because we have already consumed the input, even if `json_increment` is `None`.

Leoyzen replied:
fixed

Leoyzen committed:
fix: Update _streamed_raw_length before early return in GLM MoE detectors

Fix critical bug where _streamed_raw_length was not updated when
state machine returns empty json_increment, causing input reprocessing.

Changes:
- glm47_moe_detector.py: Move _streamed_raw_length update before
  'if not json_increment' check in _process_arguments_streaming()
- glm4_moe_detector.py: Move _streamed_raw_length update before
  'if json_increment' check in streaming logic
- Add 4 test cases to verify the fix and prevent regression

The bug occurred because even when json_increment is empty (state
machine buffering), the raw input has been consumed and should be
marked as processed by updating _streamed_raw_length.

Test Results:
- test_streamed_raw_length_buffer_only_partial_tag: FAILED before fix,
  PASSED after fix (successfully reproduced and fixed the bug)
- All 26 tests in test_glm47_moe_detector.py pass after fix

JustinTong0323 committed:
Merge branch 'main' into feature/yuchen.liu/glm47-parser-fix
